axelor-sale/src/main/java/com/axelor/apps/sale/service/saleorderline/SaleOrderLineContextHelper.java

```diff
@@ -19,15 +19,71 @@
 package com.axelor.apps.sale.service.saleorderline;
 
 import com.axelor.apps.sale.db.SaleOrder;
+import com.axelor.apps.sale.db.SaleOrderLine;
+import com.axelor.apps.sale.db.repo.SaleOrderLineRepository;
+import com.axelor.inject.Beans;
 import com.axelor.rpc.Context;
 import com.axelor.utils.helpers.ContextHelper;
+import java.util.Optional;
 
 public class SaleOrderLineContextHelper {
 
   private SaleOrderLineContextHelper() {}
 
-  public static SaleOrder getSaleOrder(Context context) {
+  public static SaleOrder getSaleOrder(Context context, SaleOrderLine saleOrderLine) {
+    SaleOrder saleOrder = ContextHelper.getOriginParent(context, SaleOrder.class);
+    if (saleOrder != null) {
+      return saleOrder;
+    }
 
-    return ContextHelper.getOriginParent(context, SaleOrder.class);
+    SaleOrderLineRepository saleOrderLineRepository = Beans.get(SaleOrderLineRepository.class);
+
+    // Line is persisted and is not a subline
+    if (saleOrderLine.getId() != null) {
+      saleOrder = saleOrderLineRepository.find(saleOrderLine.getId()).getSaleOrder();
+      if (saleOrder != null) {
+        return saleOrder;
+      }
+    }
+
+    // Line is not persisted
+    SaleOrderLine saleOrderLine1 = getParentSol(context);
+    saleOrderLine1 = saleOrderLineRepository.find(saleOrderLine1.getId());
+    saleOrder = saleOrderLine1.getSaleOrder();
+    if (saleOrder == null) {
+      saleOrder = getParentOrder(saleOrderLine1);
+    }
+    return saleOrder;
+  }
+
+  protected static SaleOrderLine getParentSol(Context context) {
+    SaleOrderLine saleOrderLine = context.asType(SaleOrderLine.class);
+    Long id = saleOrderLine.getId();
```

Comment:
since this method is recursive, if this method is called and no lines are persisted, we will get a NPE here when `context.getParent()` returns null